File: urbanairship-core/src/test/java/com/urbanairship/push/notifications/AirshipNotificationProviderTest.java
/* Copyright Airship and Contributors */

package com.urbanairship.push.notifications;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;

import com.urbanairship.AirshipConfigOptions;
import com.urbanairship.BaseTestCase;
import com.urbanairship.UAirship;
import com.urbanairship.push.PushMessage;

import org.junit.Before;
import org.junit.Test;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class AirshipNotificationProviderTest extends BaseTestCase {

    private AirshipNotificationProvider provider;
    private Context context = UAirship.getApplicationContext();
    private AirshipConfigOptions configOptions;
    private PushMessage pushMessage;

    @Before
    public void setup() {
        configOptions = new AirshipConfigOptions.Builder()
                .setDevelopmentAppKey("appKey")
                .setDevelopmentAppSecret("appSecret")
                .setProductionAppSecret("appSecret")
                .setProductionAppKey("appKey")
                .setNotificationIcon(10)
                .setNotificationAccentColor(20)
                .setNotificationChannel("test_channel")
                .build();

        createChannel("test_channel");

        provider = new AirshipNotificationProvider(context, configOptions);

        Bundle extras = new Bundle();
        extras.putString(PushMessage.EXTRA_ALERT, "Test Push Alert!");
        extras.putString(PushMessage.EXTRA_PUSH_ID, "0a2027a0-1766-11e4-9db0-90e2ba287ae5");
        extras.putString(PushMessage.EXTRA_NOTIFICATION_TAG, "some-tag");

        pushMessage = new PushMessage(extras);
    }

    /**
     * Test empty alert should result in a CANCEL status.
     */
    @Test
    public void testBuildNotificationNull() {
        Bundle empty = new Bundle();
        PushMessage emptyPushMessage = new PushMessage(empty);

        NotificationArguments arguments = provider.onCreateNotificationArguments(context, emptyPushMessage);
        NotificationResult result = provider.onCreateNotification(context, arguments);
        assertEquals(NotificationResult.CANCEL, result.getStatus());
    }

    /**
     * Test creating a notification.
     */
    @Test
    public void testBuildNotification() {
        NotificationArguments arguments = provider.onCreateNotificationArguments(context, pushMessage);
        NotificationResult result = provider.onCreateNotification(context, arguments);

        assertEquals(NotificationResult.OK, result.getStatus());
        assertNotNull(result.getNotification());
    }

    /**
     * Test the defaults.
     */
    @Test
    public void testDefaults() {
        assertEquals(10, provider.getSmallIcon());
        assertEquals(20, provider.getDefaultAccentColor());
        assertEquals("test_channel", provider.getDefaultNotificationChannelId());
    }

    /**
     * Test arguments when the channel is empty.
     */
    @Test
    public void testArgumentsDefaultChannel() {
        provider.setDefaultNotificationChannelId("does not exist");
        NotificationArguments arguments = provider.onCreateNotificationArguments(context, pushMessage);

        assertEquals(NotificationProvider.DEFAULT_NOTIFICATION_CHANNEL, arguments.getNotificationChannelId());
    }

    /**
     * Test the channel will fallback to the SDK default channel if the specified channel does not exist.
     */
    @Test
    public void testArgumentsFallbackChannel() {
        NotificationArguments arguments = provider.onCreateNotificationArguments(context, pushMessage);

        assertEquals("test_channel", arguments.getNotificationChannelId());
        assertEquals(pushMessage, arguments.getMessage());
        assertEquals(AirshipNotificationProvider.TAG_NOTIFICATION_ID, arguments.getNotificationId());
        assertEquals("some-tag", arguments.getNotificationTag());
        assertFalse(arguments.getRequiresLongRunningTask());
    }

    /**
     * Test arguments when the channel is defined on the push message.
     */
    @Test
    public void testArgumentsWithChannel() {
        Bundle extras = new Bundle();
        extras.putString(PushMessage.EXTRA_ALERT, "Test Push Alert!");
        extras.putString(PushMessage.EXTRA_PUSH_ID, "0a2027a0-1766-11e4-9db0-90e2ba287ae5");
        extras.putString(PushMessage.EXTRA_NOTIFICATION_CHANNEL, "cool-channel");
        extras.putString(PushMessage.EXTRA_NOTIFICATION_TAG, "some-tag");

        createChannel("cool-channel");

        pushMessage = new PushMessage(extras);

        NotificationArguments arguments = provider.onCreateNotificationArguments(context, pushMessage);
        assertEquals("cool-channel", arguments.getNotificationChannelId());
        assertEquals(pushMessage, arguments.getMessage());
        assertEquals(AirshipNotificationProvider.TAG_NOTIFICATION_ID, arguments.getNotificationId());
        assertEquals("some-tag", arguments.getNotificationTag());
        assertFalse(arguments.getRequiresLongRunningTask());
    }

    /**
     * Test overriding the small icon from a push.
     */
    @Test
    public void testOverrideSmallIcon() {
        PushMessage mockPush = mock(PushMessage.class);
        when(mockPush.getAlert()).thenReturn("alert");
        when(mockPush.getIcon(eq(context), anyInt())).thenReturn(100);

        NotificationArguments arguments = provider.onCreateNotificationArguments(context, mockPush);
        NotificationResult result = provider.onCreateNotification(context, arguments);

        assertEquals(NotificationResult.OK, result.getStatus());
        assertEquals(100, result.getNotification().getSmallIcon().getResId());
    }

    private void createChannel(@NonNull String channelId) {
        NotificationManager manager = (NotificationManager) getApplication().getSystemService(Context.NOTIFICATION_SERVICE);
        manager.createNotificationChannel(new NotificationChannel(channelId, "test", NotificationManager.IMPORTANCE_DEFAULT));
    }

}

File: urbanairship-core/src/test/java/com/urbanairship/push/notifications/AirshipNotificationProviderTest.kt
/* Copyright Airship and Contributors */
package com.urbanairship.push.notifications

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import androidx.core.os.bundleOf
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.urbanairship.AirshipConfigOptions
import com.urbanairship.TestApplication
import com.urbanairship.Airship
import com.urbanairship.push.PushMessage
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class AirshipNotificationProviderTest {

    private var configOptions = AirshipConfigOptions.Builder()
        .setDevelopmentAppKey("appKey")
        .setDevelopmentAppSecret("appSecret")
        .setProductionAppSecret("appSecret")
        .setProductionAppKey("appKey")
        .setNotificationIcon(10)
        .setNotificationAccentColor(20)
        .setNotificationChannel("test_channel")
        .build()

    private val context = Airship.applicationContext
    private var provider = AirshipNotificationProvider(context, configOptions)

    private var defaultPushMessage = PushMessage(
        bundleOf(
            PushMessage.EXTRA_ALERT to "Test Push Alert!",
            PushMessage.EXTRA_PUSH_ID to "0a2027a0-1766-11e4-9db0-90e2ba287ae5",
            PushMessage.EXTRA_NOTIFICATION_TAG to "some-tag"
        )
    )

    @Before
    fun setup() {
        createChannel("test_channel")
    }

    /**
     * Test empty alert should result in a CANCEL status.
     */
    @Test
    fun testBuildNotificationNull() {
        val emptyPushMessage = PushMessage(bundleOf())

        val arguments = provider.onCreateNotificationArguments(context, emptyPushMessage)
        val result = provider.onCreateNotification(context, arguments)
        Assert.assertEquals(NotificationResult.Status.CANCEL, result.status)
    }

    /**
     * Test creating a notification.
     */
    @Test
    fun testBuildNotification() {
        val arguments = provider.onCreateNotificationArguments(context, defaultPushMessage)
        val result = provider.onCreateNotification(context, arguments)

        Assert.assertEquals(NotificationResult.Status.OK, result.status)
        Assert.assertNotNull(result.notification)
    }

    /**
     * Test the defaults.
     */
    @Test
    fun testDefaults() {
        Assert.assertEquals(10, provider.smallIcon)
        Assert.assertEquals(20, provider.defaultAccentColor)
        Assert.assertEquals("test_channel", provider.defaultNotificationChannelId)
    }

    /**
     * Test arguments when the channel is empty.
     */
    @Test
    fun testArgumentsDefaultChannel() {
        provider.defaultNotificationChannelId = "does not exist"
        val arguments = provider.onCreateNotificationArguments(context, defaultPushMessage)

        Assert.assertEquals(
            NotificationProvider.DEFAULT_NOTIFICATION_CHANNEL, arguments.notificationChannelId
        )
    }

    /**
     * Test the channel will fallback to the SDK default channel if the specified channel does not exist.
     */
    @Test
    fun testArgumentsFallbackChannel() {
        val arguments = provider.onCreateNotificationArguments(context, defaultPushMessage)

        Assert.assertEquals("test_channel", arguments.notificationChannelId)
        Assert.assertEquals(defaultPushMessage, arguments.message)
        Assert.assertEquals(
            AirshipNotificationProvider.TAG_NOTIFICATION_ID, arguments.notificationId
        )
        Assert.assertEquals("some-tag", arguments.notificationTag)
        Assert.assertFalse(arguments.requiresLongRunningTask)
    }

    /**
     * Test arguments when the channel is defined on the push message.
     */
    @Test
    fun testArgumentsWithChannel() {
        createChannel("cool-channel")

        val pushMessage = PushMessage(
            bundleOf(
                PushMessage.EXTRA_ALERT to "Test Push Alert!",
                PushMessage.EXTRA_PUSH_ID to "0a2027a0-1766-11e4-9db0-90e2ba287ae5",
                PushMessage.EXTRA_NOTIFICATION_CHANNEL to "cool-channel",
                PushMessage.EXTRA_NOTIFICATION_TAG to "some-tag"
            )
        )

        val arguments = provider.onCreateNotificationArguments(context, pushMessage)
        Assert.assertEquals("cool-channel", arguments.notificationChannelId)
        Assert.assertEquals(pushMessage, arguments.message)
        Assert.assertEquals(
            AirshipNotificationProvider.TAG_NOTIFICATION_ID, arguments.notificationId
        )
        Assert.assertEquals("some-tag", arguments.notificationTag)
        Assert.assertFalse(arguments.requiresLongRunningTask)
    }

    /**
     * Test overriding the small icon from a push.
     */
    @Test
    fun testOverrideSmallIcon() {
        val mockPush: PushMessage = mockk(relaxed = true) {
            every { alert } returns "alert"
            every { getIcon(context, any()) } returns 100
        }

        val arguments = provider.onCreateNotificationArguments(context, mockPush)
        val result = provider.onCreateNotification(context, arguments)

        Assert.assertEquals(NotificationResult.Status.OK, result.status)
        Assert.assertEquals(100, result.notification?.smallIcon?.resId)
    }

    private fun createChannel(channelId: String) {
        val manager = TestApplication.getApplication()
            .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        manager.createNotificationChannel(
            NotificationChannel(channelId, "test", NotificationManager.IMPORTANCE_DEFAULT)
        )
    }
}
